x-pack/plugins/fleet/public/applications/fleet/sections/agents/agent_list_page/components/agent_upgrade_status.tsx

```diff
@@ -0,0 +1,222 @@
+    case 'UPG_RESTARTING':
+      return {
+        Badge: (
+          <EuiBadge color="warning">
+            <FormattedMessage
+              id="xpack.fleet.agentUpgradeStatusBadge.upgradeRestarting"
+              defaultMessage="Upgrade restarting"
+            />
+          </EuiBadge>
+        ),
+        TooltipText: (
+          <FormattedMessage
+            id="xpack.fleet.agentUpgradeStatusTooltip.upgradeRestarting"
+            defaultMessage="The agent is restarting to apply the update."
+          />
+        ),
+      };
+    case 'UPG_WATCHING':
+      return {
+        Badge: (
+          <EuiBadge color="warning">
+            <FormattedMessage
+              id="xpack.fleet.agentUpgradeStatusBadge.upgradeMonitoring"
+              defaultMessage="Upgrade monitoring"
+            />
+          </EuiBadge>
+        ),
+        TooltipText: (
+          <FormattedMessage
+            id="xpack.fleet.agentUpgradeStatusTooltip.upgradeMonitoring"
+            defaultMessage="Monitoring the new agent version for errors."
+          />
+        ),
+      };
+    case 'UPG_ROLLBACK':
+      return {
+        Badge: (
+          <EuiBadge color="danger">
+            <FormattedMessage
+              id="xpack.fleet.agentUpgradeStatusBadge.upgradeRolledBack"
+              defaultMessage="Upgrade rolled back"
+            />
+          </EuiBadge>
+        ),
+        TooltipText: (
+          <FormattedMessage
+            id="xpack.fleet.agentUpgradeStatusTooltip.upgradeRolledBack"
+            defaultMessage="Upgrade unsuccessful. Rolling back to previous version."
+          />
+        ),
+      };
+    case 'UPG_FAILED':
+      return {
+        Badge: (
+          <EuiBadge color="danger">
+            <FormattedMessage
+              id="xpack.fleet.agentUpgradeStatusBadge.upgradeFailed"
+              defaultMessage="Upgrade failed"
+            />
+          </EuiBadge>
+        ),
+        TooltipText: (
+          <FormattedMessage
+            id="xpack.fleet.agentUpgradeStatusTooltip.upgradeFailed"
+            defaultMessage="Upgrade failed: {errorMsg}."
+            values={{
+              errorMsg: agentUpgradeDetails?.metadata.error_msg,
+            }}
+          />
+        ),
+      };
+    default:
+      return null;
+  }
+}
+
+export const AgentUpgradeStatus: React.FC<{
+  agentUpgradeStartedAt?: string | null;
+  agentUpgradeDetails?: AgentUpgradeDetails;
+}> = ({ agentUpgradeStartedAt, agentUpgradeDetails }) => {
+  const status = useMemo(() => getStatusComponents(agentUpgradeDetails), [agentUpgradeDetails]);
+
+  if (agentUpgradeDetails && status) {
+    return (
+      <EuiFlexGroup gutterSize="s" alignItems="center" responsive={false}>
+        <EuiFlexItem grow={false}>{status.Badge}</EuiFlexItem>
+        <EuiFlexItem grow={false}>
+          <EuiIconTip type="iInCircle" content={status.TooltipText} color="subdued" />
+        </EuiFlexItem>
+      </EuiFlexGroup>
+    );
+  }
+
+  // TODO: check logic
```

Comment:
This is something I'm not of: we want to show this tooltip if the agent is upgrading but doesn't have upgrade details. Is this check a proper way to do this?

Reply:
Changed the check to `upgrade_started_at:* AND NOT upgraded_at:*` (thanks @juliaElastic).
@ycombinator Could you confirm this is correct?

Reply:
This predicate makes sense to me. And I think it's clever to show the tooltip in this scenario (agent is upgrading but we don't have upgrade details) as it encourages users to upgrade to 8.12.  

My only feedback would be to somehow not hardcode the `8.12` if possible because we don't know yet when the entire implementation of the upgrade details (including the Agent side of things) will be done.  I'm reasonably sure we're on track for 8.12 on the Agent side of things but if, for some odd reason, we don't make that, how can we remember to update this tooltip text? Or perhaps, to be safe, we could pull out just the code for this scenario into a separate PR that we merge at the very end, when we're absolutely sure of the version upgrade details will be available, end-to-end?

Reply:
I agree hardcoding the version number is not ideal.

Perhaps a narrow way to be safe could be to define a `minVersion` variable, currently `undefined`, and only render the tooltip when its value is set. Then, when the changes are ready on Agent side, we would have to remember to open a PR to set the value to `8.12` or another.

@juliaElastic WDYT?

Reply:
Sounds good, we could create a small gh issue and put in this sprint so we don't forget to set the variable.

Reply:
@ycombinator I've pushed this change and the other one (no download percent indication in the tooltip if 0) and opened a WIP followup PR to set the version: https://github.com/elastic/kibana/pull/168638 that could be merged after the Agent changes are ready.

Could you confirm whether we're ok to merge this PR and add a task to https://github.com/elastic/elastic-agent/issues/3119 to ping Fleet team to merge https://github.com/elastic/kibana/pull/168638 when you are ready on your end?